src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,170 @@
+    }
+
+
+    private Object[] grow() {
+        int oldCapacity = array.length;
+        double multiplicator = 1;
+        if (oldCapacity == size) {
+            multiplicator = 1.5;
+        }
+        if (oldCapacity > 0 || array != empty_arraydata) {
+            int newCapacity = (int) (oldCapacity * multiplicator + 1);
+            return array = Arrays.copyOf(array, newCapacity);
+        } else {
+            return array = (T[]) new Object[def_capacity];
+        }
+    }
+
+
+    public static Object[] listToArray(List<?> list) {
+        if (!list.isEmpty()) {
+            Object[] tempArray = new Object[list.size()];
+            for (int index = 0; index < list.size(); index++) {
+                tempArray[index] = list.get(index);
+            }
+            return tempArray;
+        }
+        return null;
+    }
+
+
     @Override
     public void add(T value) {
-
+        int index = cursor;
+        ArrayList.this.add(value, index);
+        cursor = index + 1;
     }
 
+
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            int s = size;
+            if (s == array.length) {
+                array = (T[]) grow();
+            }
+            System.arraycopy(array, index,
+                array, index + 1,
+                s - index);
+            array[index] = value;
+            size = s + 1;
+        }
     }
 
+
     @Override
     public void addAll(List<T> list) {
-
+        if (!list.isEmpty()) {
+            T[] tempArray = (T[]) listToArray(list);
+            int newCapacity = size + tempArray.length;
+            array = Arrays.copyOf(array, newCapacity);
+            System.arraycopy(tempArray, 0, array, size, tempArray.length);
+            size = newCapacity;
+        }
     }
 
+
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            return array[index];
+        }
     }
 
+
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            array[index] = value;
+        }
     }
 
+
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            int newSize = size - 1;
+            final T oldValue = get(index);
+            if (newSize > index) {
+                System.arraycopy(array, index + 1, array, index, newSize - index);
+            }
+            size = newSize;
+            array[newSize] = null;
+            return oldValue;
+        }
     }
 
+
     @Override
```

Comment:
There should be only 1 empty line between methods in a class.
```suggestion
    }

    @Override
```